app/plugin_api.go

```diff
@@ -1063,3 +1063,34 @@ func (api *PluginAPI) DeleteCommand(commandID string) error {
 
 	return nil
 }
+
+// PublishPluginClusterEvent broadcasts a plugin event to all other running instances of
+// the calling plugin.
+func (api *PluginAPI) PublishPluginClusterEvent(ev model.PluginClusterEvent,
+	opts model.PluginClusterEventSendOptions) error {
+	if api.app.Cluster() == nil {
+		return nil
+	}
+
+	msg := &model.ClusterMessage{
+		Event:            model.CLUSTER_EVENT_PLUGIN_EVENT,
+		SendType:         opts.SendType,
+		WaitForAllToSend: false,
+		Props: map[string]string{
+			"PluginID": api.id,
+			"EventID":  ev.Id,
+		},
+		Data: string(ev.Data),
+	}
+
+	// If ReceiverId is not set we broadcast to all other cluster nodes.
+	if opts.ReceiverId == "" {
+		api.app.Cluster().SendClusterMessage(msg)
+	} else {
+		if err := api.app.Cluster().SendClusterMessageToNode(opts.ReceiverId, msg); err != nil {
```

Comment:
So how does the plugin know of all the node ids in the cluster?

Reply:
I've briefly explained my reasoning in https://github.com/mattermost/mattermost-server/pull/17495#issuecomment-826705963. Each node is able to figure out its own cluster id so it can be either broadcast to others or used as part of some key in KVStore. I think that's enough for what we are planning now and we always have the option to export `GetClusterInfos()` (or some variant) to the plugin API to facilitate this even further.